src/number_theory/legendre_symbol/jacobi_symbol.lean

```diff
@@ -0,0 +1,532 @@
+such that `a = 2^e * a'` and `a'` is odd. -/
+lemma two_pow_mul_odd {a : ℕ} (ha : a ≠ 0) : ∃ e a' : ℕ, odd a' ∧ a = 2 ^ e * a' :=
+begin
+  refine rec_on_mul (λ hf, false.rec _ (hf rfl)) (λ _, ⟨0, 1, odd_one, by rw [pow_zero, mul_one]⟩)
+                    (λ p pp _, _) (λ m n hm hn hmn, _) a ha,
+  { cases pp.eq_two_or_odd' with hp₂ hp₂,
+    { exact ⟨1, 1, odd_one, by rw [hp₂, pow_one, mul_one]⟩, },
+    { exact ⟨0, p, hp₂, by rw [pow_zero, one_mul]⟩, } },
+  { rcases hm (left_ne_zero_of_mul hmn) with ⟨em, m', hmo, hmm⟩,
+    rcases hn (right_ne_zero_of_mul hmn) with ⟨en, n', hno, hnn⟩,
+    refine ⟨em + en, m' * n', odd_mul.mpr ⟨hmo, hno⟩, _⟩,
+    rw [hmm, hnn, pow_add, mul_mul_mul_comm], },
+end
+
+end nat
+
+namespace int
+
+/-- If `gcd a (m * n) ≠ 1`, then `gcd a m ≠ 1` or `gcd a n ≠ 1`. -/
+lemma gcd_ne_one_of_gcd_mul_ne_one {a : ℤ} {m n : ℕ} (h : a.gcd (m * n) ≠ 1) :
+  a.gcd m ≠ 1 ∨ a.gcd n ≠ 1 :=
+begin
+  by_contra' h₁,
+  exact h (gcd_eq_one_iff_coprime.mpr $
+    is_coprime.mul_right (gcd_eq_one_iff_coprime.mp h₁.left) $ gcd_eq_one_iff_coprime.mp h₁.right)
+end
+
+/-- If `gcd a (m * n) = 1`, then `gcd a m = 1`. -/
+lemma gcd_eq_one_of_gcd_mul_eq_one_left {a : ℤ} {m n : ℕ} (h : a.gcd (m * n) = 1) : a.gcd m = 1 :=
+nat.dvd_one.mp $ trans_rel_left _ (int.gcd_dvd_gcd_mul_right_right a m n) h
+
+/-- If `gcd a (m * n) = 1`, then `gcd a n = 1`. -/
+lemma gcd_eq_one_of_gcd_mul_eq_one_right {a : ℤ} {m n : ℕ} (h : a.gcd (m * n) = 1) : a.gcd n = 1 :=
+nat.dvd_one.mp $ trans_rel_left _ (int.gcd_dvd_gcd_mul_left_right a n m) h
+
+end int
+
+namespace zmod
+
+/-- If an integer `a` and a prime `p` satisfy `gcd a p = 1`, then `a : zmod p` is nonzero. -/
+lemma ne_zero_of_gcd_eq_one {a : ℤ} {p : ℕ} (pp : p.prime) (h : a.gcd p = 1) :
+  (a : zmod p) ≠ 0 :=
+λ hf, pp.not_dvd_one (int.coe_nat_dvd.mp $ trans_rel_left _
+       (int.dvd_gcd ((int_coe_zmod_eq_zero_iff_dvd a p).mp hf) (dvd_refl p)) (congr_arg coe h))
+
+/-- If an integer `a` and a prime `p` satisfy `gcd a p ≠ 1`, then `a : zmod p` is zero. -/
+lemma eq_zero_of_gcd_ne_one {a : ℤ} {p : ℕ} (pp : p.prime) (h : a.gcd p ≠ 1) :
+  (a : zmod p) = 0 :=
+begin
+  refine (int_coe_zmod_eq_zero_iff_dvd a p).mpr _,
+  have hl := int.gcd_dvd_left a p,
+  rwa ← (nat.prime.dvd_iff_eq pp h).mp (int.coe_nat_dvd.mp (int.gcd_dvd_right a p)) at hl,
+end
+
+/-- If `p` is a prime and `a` is an integer, then `a : zmod p` is zero if and only if
+`gcd a p ≠ 1`. -/
+lemma eq_zero_iff_gcd_ne_one {a : ℤ} {p : ℕ} [pp : fact p.prime] : (a : zmod p) = 0 ↔ a.gcd p ≠ 1 :=
+⟨λ h₁ h₂, ne_zero_of_gcd_eq_one pp.1 h₂ h₁, eq_zero_of_gcd_ne_one pp.1⟩
+
+end zmod
+
+namespace list
+
+lemma pmap_append {α β : Type*} {p : α → Prop} (f : Π (a : α), p a → β) (l₁ l₂ : list α)
+  (h : ∀ (a : α), a ∈ l₁ ++ l₂ → p a) :
+  (l₁ ++ l₂).pmap f h = l₁.pmap f (λ a ha, h a (mem_append_left l₂ ha))
+                         ++ l₂.pmap f (λ a ha, h a (mem_append_right l₁ ha)) :=
```

Comment:
I'd prefer `++` to the right of `=`, since it is part of the term on the rhs. But I think you are right that (according to the mathlib style guide) it should go at the end of the previous line.